linked_list/single_linked_list.c

```diff
@@ -0,0 +1,165 @@
+#include <stdio.h>
+#include <stdlib.h>
+#define ASSERT(x) printf("Assertion: " #x " = %d\n", (x))
+
+typedef int data_t;
+
+typedef struct node {
+  data_t data;
+  struct node *next;
+} node;
+
+#define CHECK_ALLOC(ptr)						\
+  if (!ptr) {								\
+    printf("Error allocating memory for %s in  %s: %d line", #ptr, __func__, __LINE__-1); \
+    exit(2);								\
+  }
+
+node* list_create(data_t data){
+  node* first = (node*)malloc(sizeof(node));
+  CHECK_ALLOC(first);
+  first->data = data;
+  first->next = NULL;
+  return first;
+}
+
+
+void list_free(node* first){
+  struct node* tmp;
+
+  while (first != NULL)
+    {
+      tmp = first;
+      first = first->next;
+      free(tmp);
+    }
+}
+
+void list_append_element_end(node* list, int new_element){
+  node* new_node = (node*)malloc(sizeof(node)); 
+  CHECK_ALLOC(new_node); 
+  new_node->data = new_element;
+  new_node->next = NULL;
+
+  if (list->next == NULL)
+    list->next = new_node;
+  else {
+    node* current_node = list;
+    while (current_node->next != NULL) {
+      current_node = current_node->next; 
+    }
+    current_node->next = new_node; 
+  }
+}
+
+int list_pop_end(node* list){
+  
+  node* prev_node = NULL;
+  while (list->next != NULL) {
+    prev_node = list;
+    list = list->next; 
+  } 
+  
+  if (prev_node)
+    prev_node->next = NULL;
+
+  int result = list->data; 
+  free(list);
+  return result;
+}
+
+void list_append_element_beginning(node* list, int new_element){
```

Comment:
This is not used anywhere, prepare a test for it and ... try to understand why it doesn't work as you expected.